includes/class-llms-dom-document.php

```diff
@@ -0,0 +1,160 @@
+
+	/**
+	 * Constructor
+	 *
+	 * @since [version]
+	 *
+	 * @param string $source An HTML string, either a full HTML document or a partial string.
+	 * @return void|WP_Error
+	 */
+	public function __construct( $source ) {
+
+		if ( ! class_exists( 'DOMDocument' ) ) {
+			return new WP_Error( 'llms-dom-document-missing', __( 'DOMDocument not available.', 'lifterlms' ) );
+		}
+
+		/**
+		 * Filters the convert encoding method to be used when loading the source in the DOMDocument
+		 *
+		 * @param boolean $use_mb_convert_endoding Whether or not the convert encoding method should be used when loading the source in the DOMDocument.
+		 *                                         Default is `true`. Requires `mbstring` PHP extension.
+		 */
+		$use_mb_convert_encoding = apply_filters( 'llms_dom_document_use_mb_convert_encoding', true );
+		if ( ! ( $use_mb_convert_encoding && function_exists( 'mb_convert_encoding' ) ) ) {
+			$this->load_method = 'load_with_meta_utf_fixer';
+		}
+
+		$this->source = $source;
+		$this->dom    = new DOMDocument();
+	}
+
+	/**
+	 * Load the HTML string in the DOMDocument
+	 *
+	 * This function suppresses PHP warnings that would be thrown by DOMDocument when
+	 * loading a partial string or an HTML string with errors.
+	 *
+	 * @since [version]
+	 *
+	 * @return boolean|WP_Error Returns true if the source is loaded fine.
+	 *                          Or an error object when an error is encountered during loading.
+	 */
+	public function load() {
+
+		// Don't throw or log warnings.
+		$libxml_state = libxml_use_internal_errors( true );
+
+		$this->{$this->load_method}();
+
+		// Clear and restore errors.
+		libxml_clear_errors();
+		libxml_use_internal_errors( $libxml_state );
+
+		return is_wp_error( $this->error ) && $this->error->has_errors() ? $this->error : true;
+
+	}
+
+	/**
+	 * Returns the DOMDocument
+	 *
+	 * @since [version]
+	 *
+	 * @return DOMDocument Returns an instance of DOMDocument.
+	 */
+	public function dom() {
+
+		return $this->dom;
+
+	}
+
+	/**
+	 * Load the HTML string in the DOMDocument using mb_convert_econding
+	 *
+	 * @since [version]
+	 *
+	 * @return void
+	 */
+	private function load_with_mb_convert_encoding() {
+		if ( ! $this->dom->loadHTML( mb_convert_encoding( $this->source, 'HTML-ENTITIES', 'UTF-8' ) ) ) {
+			$this->error = new WP_Error( 'llms-dom-document-error', __( 'DOMDocument XML Error encountered.', 'lifterlms' ), libxml_get_errors() );
+		}
+	}
+
+	/**
+	 * Load the HTML string in the DOMDocument using the meta ut8 fixer
+	 *
+	 * @since [version]
+	 *
+	 * @return void
+	 */
+	private function load_with_meta_utf_fixer() {
+		if ( ! $this->dom->loadHTML( $this->utf8_fixer . $this->source ) ) {
+			$this->error = new WP_Error( 'llms-dom-document-error', __( 'DOMDocument XML Error encountered.', 'lifterlms' ), libxml_get_errors() );
+		}
+
+		if ( is_wp_error( $this->error ) ) {
+			return;
+		}
```

Comment:
I think we can save checking if it's an error because it would only be an error if we don't load the HTML as expected above, right?

```suggestion
		if ( ! $this->dom->loadHTML( $this->utf8_fixer . $this->source ) ) {
			$this->error = new WP_Error( 'llms-dom-document-error', __( 'DOMDocument XML Error encountered.', 'lifterlms' ), libxml_get_errors() );
			return;
		}
```

Reply:
totally right
done in 47d61de